Makefile

```diff
@@ -306,6 +313,29 @@ $(TARGET_BIN): $(TARGET_ELF)
 	@echo "Creating BIN $(TARGET_BIN)" "$(STDOUT)"
 	$(V1) $(OBJCOPY) -O binary $< $@
 
+$(TARGET_S19): $(TARGET_ELF)
+	@echo "Creating srec/S19 $(TARGET_S19)" "$(STDOUT)"
+	$(V1) $(OBJCOPY) --output-target=srec $(TARGET_S19)
+
+$(TARGET_EXST): $(TARGET_BIN)
+	@echo "Creating EXST $(TARGET_EXST)" "$(STDOUT)"
+# Linker script should allow .bin generation from a .elf which results in a file that is the same length as the FIRMWARE_SIZE.
+# These 'dd' commands will pad a short binary to length FIRMWARE_SIZE.
+	$(V1) dd if=/dev/zero ibs=1k count=$(FIRMWARE_SIZE) of=$(TARGET_EXST)
+	$(V1) dd if=$(TARGET_BIN) of=$(TARGET_EXST) conv=notrunc
+
+	@echo "Generating MD5 hash of binary" "$(STDOUT)"
+	$(V1) openssl dgst -md5 $(TARGET_EXST) > $(TARGET_BIN).md5 
```

Comment:
What is the reason that a cryptographic hash was chosen here? If the (stated) purpose is to detect / prevent accidental data corruption then this is an extremely poor choice, as it is not actually designed to guarantee detection of accidental corruption.
Another consideration is verification of the firmware integrity by the boot loader - implementing md5 will most likely use a lot more of the scarce on-board flash, and thus compete with other desirable functionality (like FAT filesystem support to allow loading of firmware from SD card).
I think this should be changed to an algorithm that is suitable for the detection of data corruption, like block-wise CRC32.

Reply:
bootloaders in the wild already use MD5 now.

nothing stopping us from adding additional hashes though, but not as part of this PR.

wikipedia says: "Although MD5 was initially designed to be used as a cryptographic hash function, it has been found to suffer from extensive vulnerabilities. It can still be used as a checksum to verify data integrity, but only against unintentional corruption" and that is exactly the purpose i'm using it for here.

note that the H750 CPU only has support for a few different hashing algorithms, MD5 being one of them see the HASH section of the reference manual if you're interested.  only minimal flash space is required for MD5 hash verification.

Reply:
> bootloaders in the wild already use MD5 now.

This is exactly why I think that 'make it work, then make it work well' is the wrong approach when it comes to designing external facing APIs. Rather than release something badly engineered to users, and then force them through the rigamarole of having to update a bootloader that is intended to be installed as a one-off a number of times, I think we should properly design this **before** we support this hardware in Betaflight. This way only early adopters will have to re-install the boot loader when they switch from non-Betaflight to Betaflight (or decide to stay with non-Betaflight).

> nothing stopping us from adding additional hashes though, but not as part of this PR.

Ok, I must have overlooked something then. Can you please add a specification of the flash data format that points out where the information is stored that the boot loader uses to find out what type of checksum is used?

> wikipedia says

Do you have any sources that are not 'wisdom of the crowds' to back your claim that using cryptographic hashes for the purpose of bit error detection is a good practice? I only find sources recommending against this, like https://eklitzke.org/crcs-vs-hash-functions.

> note that the H750 CPU only has support for a few different hashing algorithms, MD5 being one of them see the HASH section of the reference manual if you're interested. only minimal flash space is required for MD5 hash verification.

So is support for this bound to having specific hardware? If so, do all of the MCU types that we are currently support or are potentially going to support have this hardware? If not then this is just another argument for choosing a checksumming algorithm that can be implemented in a simple form, and used on all targeted MCU types.

Reply:
> bootloaders in the wild already use MD5 now.

This is exactly why I think that 'make it work, then make it work well' is not the right approach for designing external facing APIs. There is no point in putting users through the rigamarole of updating a boot loader that was designed to not to be updated a number of times just because the API was poorly engineered when it was released. I'd rather that we solve this properly before we add support for this hardware to Betaflight, so that only early adopters will have to update the boot loader once, when they switch from using non-Betaflight to using Betaflight.

> nothing stopping us from adding additional hashes though, but not as part of this PR.

Sorry, I must have overlooked this. Can you please add a specification for the format that is used for the flash data, and point out where the information is stored that allows the boot loader to figure out what error detection algorithm was used?

> wikipedia says: 

Do you have any information to back up this 'wisdom of the crowds' claim that using a cryptographic hash to detect bit errors is an acceptable practice? All information that I am finding is pointing out that this is to be avoided, like: https://eklitzke.org/crcs-vs-hash-functions

> note that the H750 CPU only has support for a few different hashing algorithms, 

So your chosen approach requires specific hardware to work? If so, do all of the MCU types that we are supporting or potentially will be supporting actually have this hardware? If no, then this is just another point for me that we should go for a much simpler algorithm that is designed for error detection, and implement it in a portable way.

Reply:
Re: Hardware assisted MD5
Hardware hash assistance is not relevant for this application, as a hash is only computed once (or couples of times) per reboot cycle.

Reply:
Also, CRYP module is only available in H75X variants. H74Xs do not have it.

Reply:
@mikeller No, I've not got sources re crc vs hash, I'll bear this in mind next time I have to choose hash vs crc for sure though, so thanks for pointing that out.

I know this isn't much help now, but the time to offer guidance on MD5 vs CRC vs something else was in February 2019 long before the bootloader using this method was shipped.  It's a bit late now...  There were discussions in Slack regarding this at the time, your nor anyone else had objections back then, you even said "yay for hardware crypto support in a flight controller. :wink:".

The chosen approach does not require specific hardware to work.  Bootloader implementors can use whatever hash/crc method they choose as long as the current MD5 system remains unchanged.

With regards to bootloader updates it's not possible.  The H750 with a single 128K page of flash has no safe way to reflash it's internal flash in a secure way.  If you know otherwise or have ideas please contact me privately with your thoughts on that.

I'll add a commit which details the firmware memory map and MD5 requirements for you.

@jflyper yes, correct on both counts, using hardware hash or crc functionality does mean you don't need to use your own implementation of a hash/crc which a) might take more flash space, b) might be error-prone.  With hardware hash you're relying on the man-hours spent by the CPU designers, not your own.

Reply:
@jflyper:

> Re: Hardware assisted MD5
> Hardware hash assistance is not relevant for this application, as a hash is only computed once (or couples of times) per reboot cycle.

I am not so much worried about speed, but about the necessity to have to fit an implementation of MD5 into the very limited flash space if other manufacturers want to use cheaper MCUs that don't have hardware crypto support.

> Also, CRYP module is only available in H75X variants. H74Xs do not have it.

That's exactly why I am opposing an implementation that tries to force everybody who wants to use this functionality to use an MCU that has hardware crypto support.

@hydra:

As far as I remember I already stated back then that CRC32 was a better choice:

> mikeller9:12 AM
> Would using CRC32 instead make things better?

> With regards to bootloader updates it's not possible. The H750 with a single 128K page of flash has no safe way to reflash it's internal flash in a secure way. If you know otherwise or have ideas please contact me privately with your thoughts on that.

I do not quite understand that, can you please elaborate a bit? To my understanding all STM32 MCUs have a non-erasable boot loader that can be used to re-flash the flash. Why not just ship a fixed version of the boot loader, and have users use either dfu-util, or a modified version of the configurator to flash it?

Reply:
@mikeller it was a while ago, but I'm now remembering why md5 was chosen, it was mainly because of the lack of cross-platform support for a tool that could calculate CRCs, so an off-the-shelf solution (openssl) was chosen.  The alternative, amazingly in 2019, appeared to be creating our own CRC program and adding it, and maintaining it forever more.  That of course requires that the user has a working host-compiler available (as opposed to just a cross-compiler).

Reply:
Commit added to document the EXST firmware format.

Reply:
@hydra:

> it was a while ago, but I'm now remembering why md5 was chosen, it was mainly because of the lack of cross-platform support for a tool that could calculate CRCs, so an off-the-shelf solution (openssl) was chosen. 

Who 'chose' this?

And who was consulted on availability of tools?

```
mikeller@pc:~$ man crc32 | head
CRC32(1)                    General Commands Manual                   CRC32(1)

NAME
       crc32 - compute CRC-32 checksums for the given files

SYNOPSIS
       crc32 filename [ filename ... ]

DESCRIPTION
       crc32  is a simple utility that calculates the CRC-32 checksum for each
```

Reply:
The specification for the EXST format looks mostly sound. The requirement for the reserved bytes to be all 0xaa seems to be a bit unreasonable - this will make it impossible to ever use these bytes without breaking compatibility with any boot loader that expects the specification to be met - this should be removed.

Can I also propose that we allocate the first of the reserved bytes (or the last one, as it is closest to the checksum) as an indicator of what checksum type is used - this will pave the way to use a less problematic algorithm when moving to Universal Targets.